### ISSUE TITLE
Pythonic Specification Generation

### ISSUE THREAD

**Issue by kratsg** [diana-hep/pyhf#106]
I strongly suggest we keep the most stringent JSON spec in #105 as the only JSON-schema we have. A "lite JSON" would probably confuse newcomers and those wishing to use an API. Instead, a more user-friendly, pythonic generation can be built. Something like the following is certainly possible [inspired by how [constructions](http://construct.readthedocs.io/en/latest/) does it]

```python
NPs = [NormSys("JES1"), HistoSys("JES2")]
sample = Sample(
    NormSys("JES1") / Data(.....),
    HistoSys("JES2") / Data(....)
)
chan = "singlechannel" / Channel( "signal" / sample)
```

then doing something like `JSON.dump(chan)` would work out of the box, as you can define how to serialize such an object. `chan` can implement `vars(chan)` which returns the simple python structure that can be passed into hfpdf -- similar to how `argparse::Namespace` does it.

Caveat: division does not need to be done. One could just as easily do `NormSys("JES1").Data("....")` and so on.

**Comment by paulgessinger** [comment]
I was thinking about this. 

[Pydantic](https://pydantic-docs.helpmanual.io) has the ability to export JSON schema from its model classes, and I believe codegen from JSON schema is also possible. One possibility here would be to codegen pydantic models from the hand written JSON schema. These pydantic models could then essentially be used as a python model building API, and maybe even be used as the source-of-truth for the schema itself.

**Comment by lukasheinrich** [comment]
thanks @paulgessinger  - certainly also something that @alexander-held has thought about

**Comment by lukasheinrich** [comment]
one could imagine a similar workflow to keras

```
from pyhf.build import histosys, normsys, model ...
m = model()
m.add(histosys(...))
pdf = m.compile()
pdf.fit(data)
```

which brings it full circle to MakeModelAndMeasurementsFast :)

**Comment by paulgessinger** [comment]
Yeah, that looks what I had in mind. 'Rendering' pydantic models to dicts which are consumed by `pyhf.Model` is most likely the simplest way to go.

**Comment by lukasheinrich** [comment]
I wonder if this is something that can be decoupled from core pyhf with  the JSON schema serving as the data exchange, (either as `pyhf.contrib` or in `cabinetry` or as a separate project)

**Comment by alexander-held** [comment]
@lukasheinrich's example looks nice and could clean up the workspace building in `cabinetry` a bit. Currently it is like this

```python
if sample_affected_by_modifier(sample, NormFactor):
    modifiers.append({"data": None, "name": NormFactor["Name"], "type": "normfactor"})
``` 

and could be more like this (though maybe it's cleaner to follow the workspace structure and add modifiers to samples than to a model)

```python
model.add(normfactor(NormFactor["Name"], get_affected_samples())
```

Factoring out the names of the keys and the exact format would also mean that things are stable should the workspace format change, since that API used to build the workspace would change at the same time.

The barrier to write a pyhf workspace feels reasonably low, and given that there really aren't a lot of different modifiers and elements to the workspace, a model building API could probably be quite lightweight. On the other hand, it's also not too hard for people to implement it themselves since they're writing something against a rather stable and well-defined workspace specification.

**Comment by kratsg** [comment]
One could consider `marshmallow` as well.

**Comment by lukasheinrich** [comment]
reviving this based on some discussions with @pfackeldey @kratsg @matthewfeickert
